### PR TITLE
STP < Refining Scoring for positions

### PR DESCRIPTION
**Pull request by mmlievenseRTT**
### Summary
Generalized position computations for a more modular functions to be used in play creation. Made the function more efficient and split up the computations and evaluations from each other. Added functionality where positions are only calculated a single time and then saved for the rest of the tick.

### Related issues
When picking a new play to change to, the important positions of the active roles will be computed and evaluated. Before the information of each computation was not saved and it would do the same computations for each alike play. Now the base score of each position is saved in a map where the weights for a particular profile (combination of weights) is used.

The efficiency and effectiveness of the position evaluations have been increased.

### CODE REVIEW

File: include/roboteam_ai/stp/constants/GeneralizationConstants.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    /**
+     * Generalized Grids
+     */
+    inline static Grid gridRightTop = Grid(3, 0, 2.5, 2.2, 5, 5);
+    inline static Grid gridRightMid = Grid(3, -1.5, 2.5, 3, 5, 5);
+    inline static Grid gridRightBot = Grid(3, -2.5, 2.5, 2.2, 5, 5);
+    inline static Grid gridMidFieldTop = Grid(-1, 0, 2, 2.2, 5, 5);
+    inline static Grid gridMidFieldMid = Grid(-1, -1.5, 2, 3, 5, 5);
+    inline static Grid gridMidFieldBot = Grid(-1, -2.5, 2, 2.2, 5, 5);
```
Comment (Yuhanun) [inline include/roboteam_ai/stp/constants/GeneralizationConstants.h]: constexpr?

File: include/roboteam_ai/stp/constants/GeneralizationConstants.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+#include <roboteam_utils/Grid.h>
+
+namespace rtt::ai::stp::gen {
+    /**
+     * Struct that is used to store computations made with this module
+     */
+    struct PositionScores {
+        std::optional<double> scoreOpen;
+        std::optional<double> scoreLineOfSight;
+        std::optional<double> scoreGoalShot;
```
Comment (Yuhanun) [inline include/roboteam_ai/stp/constants/GeneralizationConstants.h:25]: I have no clue what these 3 mean

Comment (Yuhanun) [inline include/roboteam_ai/stp/constants/GeneralizationConstants.h:25]: I have no clue what these 3 mean

File: include/roboteam_ai/stp/constants/GeneralizationConstants.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    };
+
+    /**
+     * Struct of a position profile
+     */
+    struct ScoreProfile{
+        double weightOpen;
+        double weightLineOfSight;
+        double weightGoalShot;
```
Comment (Yuhanun) [inline include/roboteam_ai/stp/constants/GeneralizationConstants.h:40]: Same for these

File: include/roboteam_ai/stp/evaluations/position/GoalShotEvaluation.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
+//
+// Created by maxl on 23-03-21.
+//
+
+#ifndef RTT_GOALSHOTEVALUATION_H
+#define RTT_GOALSHOTEVALUATION_H
+
+#include <NFParam/Param.h>
```
Comment (Yuhanun) [inline include/roboteam_ai/stp/evaluations/position/GoalShotEvaluation.h:8]: ? do we really still use this?

File: src/stp/computations/PassComputations.cpp [hunk 5]
```diff
@@ -16,17 +16,17 @@ namespace rtt::ai::stp::computations {
...
+//            return std::max_element(positions.begin(), positions.end(),
+//                                    [](PositionComputations::PositionEvaluation &left,
+//                                       PositionComputations::PositionEvaluation &right) {
+//                                        return left.score < right.score;
+//                                    })->position;
+//        } else {
+//            RTT_DEBUG("There were no possible locations to pass to.");
+//            return {};
+//        }
+//    }
```
Comment (Yuhanun) [inline src/stp/computations/PassComputations.cpp:31]: Why was this commented? can it be removed?

File: src/stp/computations/PositionComputations.cpp [hunk 6]
```diff
@@ -1,167 +1,97 @@
...
 
-    PositionComputations::PositionEvaluation PositionComputations::determineBestLineOfSightPosition(const Grid &searchGrid, const rtt::world::Field &field,
-                                                                                      rtt::world::World *world) {
-        double bestScore = 0;
-        Vector2 bestPosition{};
-
-        // From given grid, check the best point
+    gen::ScoredPosition PositionComputations::getPosition(std::optional<rtt::Vector2> currentPosition, const Grid &searchGrid, gen::ScoreProfile profile, const world::Field &field, world::World *world) {
+        gen::ScoredPosition bestPosition;
+        (currentPosition.has_value()) ? bestPosition = scorePosition(currentPosition.value(), profile, field, world, 2) : bestPosition = {{0,0},0};
```
Comment (Yuhanun) [inline src/stp/computations/PositionComputations.cpp:21]:
```suggestion
        bestPosition = (currentPosition.has_value()) ? scorePosition(currentPosition.value(), profile, field, world, 2) : {{0,0},0};
```

File: src/stp/evaluations/position/GoalShotEvaluation.cpp [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+                 * Closer is better, caps at far distances
+                 * -> Max score till 2 meters, linear from 2 to 7 meters.
+                 */
+                 * std::max(std::min(5.0,(7-dist))/5,0.0)
+                 /**
+                  * Angle, Linear -> Favor in front
+                  * In front in better
+                  * -> Max score till 30 degrees, linear from 30 to 90.
+                  */
+                  * std::max(std::min(M_PI_2-angle,M_PI/6)/M_PI/6,0.0);
```
Comment (Yuhanun) [inline src/stp/evaluations/position/GoalShotEvaluation.cpp]: This is super confusing, the * seems liek a part of the comments lol

File: src/stp/plays/offensive/AttackingPass.cpp [hunk 8]
```diff
@@ -162,48 +135,49 @@ bool AttackingPass::shouldRoleSkipEndTactic() { return false; }
...
+//            return passLine.contains(bot->getPos());
+//        })) {
+//        stpInfos["passer"].setKickOrChip(KickOrChip::CHIP);
+//    } else {
+//        stpInfos["passer"].setKickOrChip(KickOrChip::KICK);
+//    }
+//    // Passer
+//    stpInfos["passer"].setPositionToShootAt(passingPosition);
+//    stpInfos["passer"].setShotType(ShotType::TARGET);
+//}
```
Comment (Yuhanun) [inline src/stp/plays/offensive/AttackingPass.cpp:178]: can this be removed?

**Review by Yuhanun — CHANGES_REQUESTED** [review]

File: include/roboteam_ai/stp/evaluations/position/GoalShotEvaluation.h (hunk 4 above)
Comment (mmlievenseRTT) [reply include/roboteam_ai/stp/evaluations/position/GoalShotEvaluation.h:8]: This one in particular not. When re-adding the plays into the AI (next step) a more of the old invariants will pop out as obsolete, I'll remove them then.

File: src/stp/computations/PassComputations.cpp (hunk 5 above)
Comment (mmlievenseRTT) [reply src/stp/computations/PassComputations.cpp:31]: Sorry, forgot the comment, The code cant work in this play and i needed the play for testing. The code iwll be used in a new play and therefore i didnt remove it just yet.

File: src/stp/computations/PositionComputations.cpp (hunk 6 above)
Comment (mmlievenseRTT) [reply src/stp/computations/PositionComputations.cpp:21]: ->Initializer list cannot be used on the right hand side of operator ':'
that's the reason i did it ugly :(

File: src/stp/evaluations/position/GoalShotEvaluation.cpp (hunk 7 above)
Comment (mmlievenseRTT) [reply src/stp/evaluations/position/GoalShotEvaluation.cpp]: I can fix it with indenting,

File: src/stp/plays/offensive/AttackingPass.cpp (hunk 8 above)
Comment (mmlievenseRTT) [reply src/stp/plays/offensive/AttackingPass.cpp:178]: Same as the other block, need to code later, but needs to be disabled to test

File: include/roboteam_ai/stp/constants/GeneralizationConstants.h (hunk 1 above)
Comment (mmlievenseRTT) [reply include/roboteam_ai/stp/constants/GeneralizationConstants.h]: the profiles can be constexpr. Constructing the Grid with constexpr requires such a constructor. Is there a net method of adding such a constructor?

File: include/roboteam_ai/stp/StpInfo.h [hunk 9]
```diff
@@ -50,6 +51,7 @@ struct StpInfo {
 
     const std::optional<Vector2>& getPositionToMoveTo() const { return positionToMoveTo; }
     void setPositionToMoveTo(const std::optional<Vector2>& position) { this->positionToMoveTo = position; }
+    void setPositionToMoveTo(const std::optional<gen::ScoredPosition>& scoredPosition) { setRoleScore(scoredPosition->score); setPositionToMoveTo(scoredPosition->position);}
```
Comment (rolfvdhulst) [inline include/roboteam_ai/stp/StpInfo.h]: Separate into two lines, it's two  function calls.

File: include/roboteam_ai/stp/computations/PassComputations.h [hunk 10]
```diff
@@ -22,12 +22,12 @@ namespace rtt::ai::stp::computations {
...
-         * @param positions std::pair of the Vector2 in first and the given score in second
-         * @return Vector2 with highest scoring score.
-         */
-        static Vector2 determineBestPosForPass(std::vector<PositionComputations::PositionEvaluation>& positions);
+//        /**
+//         * Return the position (.first) with the highest score (.second) in the positions vector
+//         * @param positions std::pair of the Vector2 in first and the given score in second
+//         * @return Vector2 with highest scoring score.
+//         */
+//        static Vector2 determineBestPosForPass(std::vector<PositionComputations::PositionEvaluation>& positions);
```
Comment (rolfvdhulst) [inline include/roboteam_ai/stp/computations/PassComputations.h]: double comment? Just delete the function?

File: include/roboteam_ai/stp/plays/offensive/AttackingPass.h [hunk 11]
```diff
@@ -68,7 +68,7 @@ class AttackingPass : public Play {
      * Receivers will get positions to receive at, of which one will actually intercept the ball once it is close enough
      * @param ball
      */
-    void calculateInfoForPass(const world::ball::Ball* ball) noexcept;
+    //NEW PLAY -> void calculateInfoForPass(const world::ball::Ball* ball) noexcept;
```
Comment (rolfvdhulst) [inline include/roboteam_ai/stp/plays/offensive/AttackingPass.h:71]: ?

File: src/stp/evaluations/position/OpennessEvaluation.cpp [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+uint8_t OpennessEvaluation::metricCheck(std::vector<double>& enemyDistances) const noexcept {
+    double evalScore = 0.0;
+    /**
+     * Limit the ranges from 2.5 to 0.5. [y = 1/x - 0.4]
+     * (0,1.6)  |XX
+     *          |  X
+     *          |   X
+     *          |    XX
+     *          |      XXX
```
Comment (rolfvdhulst) [inline src/stp/evaluations/position/OpennessEvaluation.cpp:33]: really like these drawings, they help a lot.

File: src/stp/plays/contested/GetBallPossession.cpp [hunk 13]
```diff
@@ -77,26 +80,14 @@ void GetBallPossession::calculateInfoForRoles() noexcept {
...
-        stpInfos["waller_1"].setPositionToMoveTo(wallPositions.at(1));
-        stpInfos["waller_2"].setPositionToMoveTo(wallPositions.at(2));
-    } else {
-        stpInfos["waller_0"].setPositionToMoveTo(Vector2(0,0));
-        stpInfos["waller_1"].setPositionToMoveTo(Vector2(0,0.2));
-        stpInfos["waller_2"].setPositionToMoveTo(Vector2(0,-0.2));
-    }
+    stpInfos["midfielder_0"].setPositionToMoveTo(PositionComputations::getPosition(stpInfos["midfielder_0"].getPositionToMoveTo(),gen::gridMidFieldBot, gen::SafePosition, field, world));
+    stpInfos["midfielder_1"].setPositionToMoveTo(PositionComputations::getPosition(stpInfos["midfielder_1"].getPositionToMoveTo(),gen::gridMidFieldMid, gen::SafePosition, field, world));
+    stpInfos["midfielder_2"].setPositionToMoveTo(PositionComputations::getPosition(stpInfos["midfielder_2"].getPositionToMoveTo(),gen::gridMidFieldTop, gen::SafePosition, field, world));
```
Comment (rolfvdhulst) [inline src/stp/plays/contested/GetBallPossession.cpp]: These lines are really long.

File: src/stp/plays/offensive/AttackingPass.cpp [hunk 14]
```diff
@@ -113,7 +94,7 @@ void AttackingPass::calculateInfoForRoles() noexcept {
 
     /// Passer and receivers
     // calculate all info necessary to execute a pass
-    calculateInfoForPass(ball);
+    //calculateInfoForPass(ball);
```
Comment (rolfvdhulst) [inline src/stp/plays/offensive/AttackingPass.cpp:97]: ?

File: src/stp/roles/active/Passer.cpp [hunk 15]
```diff
@@ -11,6 +11,6 @@ namespace rtt::ai::stp::role {
 
 Passer::Passer(std::string name) : Role(std::move(name)) {
     // create state machine and initializes the first state
-    robotTactics = collections::state_machine<Tactic, Status, StpInfo>{tactic::GetBall(), tactic::ShootAtPos()};
+    robotTactics = collections::state_machine<Tactic, Status, StpInfo>{tactic::GetBall()}; //, tactic::ShootAtPos()};
```
Comment (rolfvdhulst) [inline src/stp/roles/active/Passer.cpp:14]: What does this comment do/tell you? Unclear.

**Review by rolfvdhulst — COMMENTED** [review]
Some things here and there. Am not really a fan of commenting code out this much :/

**Pull request merged**